migrator/src/main/java/com/google/phonenumbers/migrator/MigrationJob.java

```diff
@@ -0,0 +1,160 @@
+package com.google.phonenumbers.migrator;
+
+import com.google.i18n.phonenumbers.metadata.RangeSpecification;
+import com.google.i18n.phonenumbers.metadata.RangeTree;
+import com.google.i18n.phonenumbers.metadata.i18n.PhoneRegion;
+import com.google.i18n.phonenumbers.metadata.table.CsvTable;
+import com.google.i18n.phonenumbers.metadata.table.RangeKey;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Scanner;
+
+/**
+ * Represents a migration operation for a given region where each {@link MigrationJob} contains
+ * a {@link RangeTree} of numbers to be migrated as well as the {@link CsvTable} which will hold the
+ * available recipes that can be performed on the range. Only recipes from the given regionCode will
+ * be used.
+ */
+public final class MigrationJob {
+
+  private final static String DEFAULT_RECIPES_PATH = "../recipes.csv";
+
+  private final CsvTable<RangeKey> recipesTable;
+  private final RangeTree numberRange;
+  private final PhoneRegion regionCode;
+
+  public CsvTable<RangeKey> getRecipesTable() {
+    return recipesTable;
+  }
+
+  public RangeTree getNumberRange() {
+    return numberRange;
+  }
+
+  public PhoneRegion getRegionCode() {
+    return regionCode;
+  }
+
+  /**
+   * Returns the sub range of numbers within numberRange that can be migrated using any recipe from
+   * the {@link CsvTable} recipesTable that matches the specified BCP-47 region code. This method will
+   * not perform migrations and as a result, the validity of migrations using the given recipesTable
+   * cannot be verified.
+   */
+  public RangeTree getAllMigratableNumbers() {
+    return RecipesTableSchema.toRangeTable(recipesTable)
+        .getRanges(RecipesTableSchema.REGION_CODE, regionCode)
+        .intersect(numberRange);
+  }
+
+  /**
+   * Returns the sub range of numbers within numberRange that can be migrated using the given recipe.
+   * This method will not perform migrations and as a result, the validity of migrations using the
+   * given recipe cannot be verified.
+   *
+   * @param recipeKey: the key of the recipe that is being checked
+   * @throws IllegalArgumentException if there is no row in the recipesTable with the given recipeKey
+   */
+  public RangeTree getMigratableNumbers(RangeKey recipeKey) {
+    if (!recipesTable.containsRow(recipeKey)) {
+      throw new IllegalArgumentException(
+          recipeKey + " does not match any recipe row in the given recipes table");
+    }
+    return recipeKey.asRangeTree().intersect(numberRange);
+  }
+
+  /**
+   * Removes spaces and '+' characters expected in E.164 numbers and returns the
```

Comment:
Could also be '()' and '-'. It might be useful to remove any non-digit characters.

Reply:
I think it may be beneficial to be lenient and letting known number special characters be parsed (e.g. '()', '-', ' ') as long as when stripped it becomes a valid E.164 number. I will look into storing original inputs with data because as you said, people may key off these values. Thanks